README.rst
.. contents::

Introduction
============

Templates for OpenBSD stuff

**mr.bobby** templates : http://mrbobby.readthedocs.org/en/latest/

+ carp_iface : provide carp ifaces master and slave which muste be based on a vlan interface

Documentation
=============

http://bobbytemplatesthd.readthedocs.org/

Tests
=====

bobbytemplates.thd is continuously 

+ tested on Travis |travisstatus|_ 

+ coverage tracked on coveralls.io |coveralls|_.


.. |travisstatus| image:: https://api.travis-ci.org/jpcw/bobbytemplates.thd.png?branch=master
.. _travisstatus:  http://travis-ci.org/jpcw/bobbytemplates.thd


.. |coveralls| image:: https://coveralls.io/repos/jpcw/bobbytemplates.thd/badge.png
.. _coveralls: https://coveralls.io/r/jpcw/bobbytemplates.thd



docs/CHANGES.rst
Changelog
=========

0.1-dev (unreleased)
------------------------

- Jean-Philippe Camguilhem <jpcw@camguilhem.net>


src/bobbytemplates/thd/tests/test_hooks.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Doc here.
"""

__docformat__ = 'restructuredtext en'

import os
#import shutil
#import tempfile
from unittest import TestCase


class DummyConfigurator(object):
    def __init__(self,
                 defaults=None,
                 bobbyconfig=None,
                 templateconfig=None,
                 variables=None,
                 quiet=False):
        self.defaults = defaults or {}
        self.bobbyconfig = bobbyconfig or {}
        self.variables = variables or {'physdev': 'em0'}
        self.quiet = quiet
        self.templateconfig = templateconfig or {}


class ToolsTest(TestCase):

    def test_within_intervall(self):
        from mrbobby.bobbyexceptions import ValidationError
        from ..hooks import within_intervall
        self.assertRaises(ValidationError, within_intervall, 'O', 'test')
        self.assertRaises(ValidationError, within_intervall, '256', 'test')
        self.assertEquals(within_intervall('200', 'test'), '200')

    def test_is_a_network_address(self):
        from ..hooks import is_a_network_address
        self.assertTrue(is_a_network_address('192.168.0.1/32'))
        self.assertFalse(is_a_network_address('192.168.0.1/31'))


class Post_ask_q_carp_iface_cidrTest(TestCase):

    def call_FUT(self, answer, configurator=None, question=None):
        from ..hooks import post_ask_q_carp_iface_cidr
        return post_ask_q_carp_iface_cidr(DummyConfigurator(), question,
                                          answer)

    def test_post_ask_q_carp_iface_cidr(self):
        from mrbobby.bobbyexceptions import ValidationError
        self.assertRaises(ValidationError, self.call_FUT, 'test')
        self.assertRaises(ValidationError, self.call_FUT, 'test/32')
        self.assertRaises(ValidationError, self.call_FUT, '192.168.0.3/30')
        self.assertRaises(ValidationError, self.call_FUT, '192.168.0.0/31')
        self.assertEquals(self.call_FUT('192.168.0.1/30'), '192.168.0.1/30')


class Post_ask_q_carp_iface_advskewTest(TestCase):

    def call_FUT(self, answer, configurator=None, question=None):
        from ..hooks import post_ask_q_carp_iface_advskew
        return post_ask_q_carp_iface_advskew(DummyConfigurator(), question,
                                             answer)

    def test_post_ask_q_carp_iface_advskew(self):
        from mrbobby.bobbyexceptions import ValidationError
        self.assertRaises(ValidationError, self.call_FUT, 'test')
        self.assertRaises(ValidationError, self.call_FUT, '0')
        self.assertRaises(ValidationError, self.call_FUT, '255')
        self.assertEquals(self.call_FUT('254'), '254')


class Post_ask_q_carp_iface_vlanTest(TestCase):

    def call_FUT(self, answer, configurator=None, question=None):
        from ..hooks import post_ask_q_carp_iface_vlan
        return post_ask_q_carp_iface_vlan(DummyConfigurator(), question,
                                          answer)

    def test_post_ask_q_carp_iface_vlan(self):
        from mrbobby.bobbyexceptions import ValidationError
        self.assertRaises(ValidationError, self.call_FUT, '4095')
        self.assertEquals(self.call_FUT('254'), '254')
        self.assertFalse(self.call_FUT('0'))


class Post_ask_q_carp_iface_descriptionTest(TestCase):

    def call_FUT(self, answer, configurator=None, question=None):
        from ..hooks import post_ask_q_carp_iface_description
        return post_ask_q_carp_iface_description(DummyConfigurator(), question,
                                                 answer)

    def test_post_ask_q_carp_iface_description(self):
        from string import ascii_letters
        self.assertEquals(len(self.call_FUT(ascii_letters)), 48)
        self.assertEquals(self.call_FUT(' test '), 'test')


class Post_ask_q_carp_iface_vhidTest(TestCase):
    def setUp(self):
        import bobbytemplates.thd
        self.fs_templates = os.path.abspath(
            os.path.join(os.path.dirname(bobbytemplates.thd.__file__),
                         'carp_iface'))

    def call_FUT(self, answer, configurator=None, question=None):
        from mrbobby.configurator import Configurator
        from ..hooks import post_ask_q_carp_iface_vhid
        return post_ask_q_carp_iface_vhid(Configurator(self.fs_templates, '',
                                                       {}), question, answer)

    def test_post_ask_q_carp_iface_vhid(self):
        from mrbobby.bobbyexceptions import ValidationError
        self.assertRaises(ValidationError, self.call_FUT, 'test')
        self.assertRaises(ValidationError, self.call_FUT, '0')
        self.assertEquals(self.call_FUT('255'), '255')

# vim:set et sts=4 ts=4 tw=80:


src/bobbytemplates/thd/hooks.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Render bobbytemplates.jpcw hooks.
"""

__docformat__ = 'restructuredtext en'


from IPy import IP
from mrbobby.bobbyexceptions import ValidationError


def within_intervall(str_value, name, lower_bound=1, upper_bound=255):
    """Check if int(answer) withoin intervall."""
    answer = str_value.strip()

    if answer.isdigit():
        value = int(answer)
        if value not in range(lower_bound, upper_bound + 1):
            msg = "{0} Acceptable values for {1} are from {2} to {3}"
            raise ValidationError(msg.format(answer, name, lower_bound,
                                  upper_bound))

    else:
        raise ValidationError("{0} is not a valid number".format(answer))

    return answer


def is_a_network_address(address):
    """Check if address is a network address."""
    try:
        net = IP(address)
        return bool(net)
    except:
        return False


def post_ask_q_carp_iface_cidr(configurator, question, answer):
    """Check subnet answer."""
    answer = answer.strip()
    net_vars = {}
    subnet = None

    try:
        if answer.count('/') != 1:
            msg = "'{0}' Plz input a cidr subnet notation here"
            raise ValidationError(msg.format(answer))
        ip, netmask = answer.split('/')
        subnet = IP(answer, make_net=True)
        if netmask not in ['32', '128']:
            # we accept network ip only with /32 or /128
            if is_a_network_address(answer):
                msg = "'{0}': oO you provided a network address,"\
                      " within not in [/32 or /128] !"
                raise ValidationError(msg.format(answer))
        net_vars['subnet'] = '%s/%s' % (str(subnet.net()), netmask)
        net_vars['netmask'] = str(subnet.netmask())
        net_vars['broadcast'] = str(subnet.broadcast())
        net_vars['ip'] = ip
        if netmask not in ['32', '128', '31', '127']:
            # we accept broadcast ip only with /32, /128, /31, /127
            if net_vars['ip'] == net_vars['broadcast']:
                msg = "'{0}': oO you provided a broadcast ip within"\
                      " not in [/32, /128, /31, /127]"
                raise ValidationError(msg.format(answer))

    except ValidationError as e:
        raise(e)

    except:
        raise ValidationError("'{0}' is not a valid cidr ip".format(answer))

    configurator.variables.update(net_vars)
    return answer


def post_ask_q_carp_iface_vhid(configurator, question, answer):
    """Check vhid answer."""
    vhid = within_intervall(answer, 'vhid')
    [q for q in configurator.questions if q.name ==
     'vlan'][0].default = vhid
    return vhid


def post_ask_q_carp_iface_advskew(configurator, question, answer):
    """Check advskew answer."""
    return within_intervall(answer, 'advskew', 1, 254)


def post_ask_q_carp_iface_description(configurator, question, answer):
    """Check description answer."""
    return answer.strip()[:48]


def post_ask_q_carp_iface_vlan(configurator, question, answer):
    """Check vlan answer."""
    answer = answer.strip()
    if answer.isdigit():
        if answer != '0':
            vlanid = within_intervall(answer, 'vlan', 1, 4094)
            if vlanid:
                configurator.variables['carpdev'] = 'vlan%s' % vlanid
                vlandev = configurator.variables['physdev']
                configurator.variables['vlandev'] = vlandev
                configurator.variables['rdr_vlan'] = True
                return vlanid

        else:
            carpdev = configurator.variables['physdev']
            configurator.variables['carpdev'] = carpdev
            configurator.variables['rdr_vlan'] = False
            return False


def carp_iface_pre_render(configurator):
    """pre render stuff."""


def carp_iface_post_render(configurator):
    """post render stuff."""

# vim:set et sts=4 ts=4 tw=80:
